### PR TITLE
Allow registering custom modules in quill editor.

### PR DESCRIPTION
**Pull request by geromegrignon**
### Feature Requests
This pull request adds support for registering custom Quill modules in the `Editor` component, allowing users to extend the editor's functionality by providing custom modules. The implementation introduces a new `customModules` input property and ensures modules are registered before the editor is initialized.

Enhancements for custom Quill module registration:
* Added a new `customModules` input property to the `Editor` component, allowing users to specify custom Quill modules to be registered before initialization. (`editor.ts`)
* Implemented the `registerQuillModules()` method in the `Editor` class, which iterates through the provided custom modules and registers them with Quill during the initialization process. (`editor.ts`)
* Introduced the `QuillModule` type definition to clarify the expected shape of custom modules, supporting both function and class module formats. (`editor.types.ts`)
* Updated imports to include the new `QuillModule` type for use in the `Editor` component. (`editor.ts`)

> Migrated from `primefaces/primeng#19421` — originally by `@Bartek20` on 2026-02-22

---
*Original base: `master` @ `f11b0ce`, head: `editor-quill-register` @ `9b7aef7`*

### CODE REVIEW

**Comment by geromegrignon** [comment]
*[Originally posted by `@copilot-pull-request-reviewer[bot]` on 2026-02-22T12:20:35.000Z]*

**Review (COMMENTED)**

## Pull request overview

This PR adds an extension point to the PrimeNG `Editor` component so consumers can register custom Quill modules before the Quill editor instance is created.

**Changes:**
- Added a `QuillModule` type to describe the shape of a registerable Quill module.
- Introduced a new `customModules` input on `Editor` and a `registerQuillModules()` step during initialization.
- Updated `Editor` to import and use the new type.

### Reviewed changes

Copilot reviewed 2 out of 2 changed files in this pull request and generated 4 comments.

| File | Description |
| ---- | ----------- |
| `packages/primeng/src/types/editor/editor.types.ts` | Adds `QuillModule` type to document/standardize the expected module shape. |
| `packages/primeng/src/editor/editor.ts` | Adds `customModules` input and registers modules before creating the Quill editor instance. |





---

💡 <a href="/primefaces/primeng/new/master/.github/instructions?filename=*.instructions.md" class="Link--inTextBlock" target="_blank" rel="noopener noreferrer">Add Copilot custom instructions</a> for smarter, more guided reviews. <a href="https://docs.github.com/en/copilot/customizing-copilot/adding-repository-custom-instructions-for-github-copilot" class="Link--inTextBlock" target="_blank" rel="noopener noreferrer">Learn how to get started</a>.

**Comment by geromegrignon** [comment]
*[Originally posted by `@copilot-pull-request-reviewer[bot]` on 2026-02-22T13:41:45.000Z]*

**Review (COMMENTED)**

## Pull request overview

Copilot reviewed 3 out of 3 changed files in this pull request and generated 2 comments.





---

💡 <a href="/primefaces/primeng/new/master/.github/instructions?filename=*.instructions.md" class="Link--inTextBlock" target="_blank" rel="noopener noreferrer">Add Copilot custom instructions</a> for smarter, more guided reviews. <a href="https://docs.github.com/en/copilot/customizing-copilot/adding-repository-custom-instructions-for-github-copilot" class="Link--inTextBlock" target="_blank" rel="noopener noreferrer">Learn how to get started</a>.

**Comment by geromegrignon** [comment]
*[Originally posted by `@copilot-pull-request-reviewer[bot]` on 2026-02-23T14:40:06.000Z]*

**Review (COMMENTED)**

## Pull request overview

Copilot reviewed 3 out of 3 changed files in this pull request and generated 4 comments.





---

💡 <a href="/primefaces/primeng/new/master/.github/instructions?filename=*.instructions.md" class="Link--inTextBlock" target="_blank" rel="noopener noreferrer">Add Copilot custom instructions</a> for smarter, more guided reviews. <a href="https://docs.github.com/en/copilot/customizing-copilot/adding-repository-custom-instructions-for-github-copilot" class="Link--inTextBlock" target="_blank" rel="noopener noreferrer">Learn how to get started</a>.

**Comment by geromegrignon** [comment]
*[Originally posted by `@Copilot` on 2026-02-22T12:20:34.000Z]*

*[Inline review on `packages/primeng/src/editor/editor.ts` could not be recreated]*

`this.dynamicQuill.registered(...)` is not a Quill API and will throw at runtime. Quill exposes a static `register(...)` method (e.g. `Quill.register(path, def)`), so this should call `this.dynamicQuill.register(...)` (and use the appropriate module path, like `modules/<name>`, per Quill docs).
```suggestion
            this.dynamicQuill.register(`modules/${moduleName}`, moduleInstance);
```

**Comment by geromegrignon** [comment]
*[Originally posted by `@Copilot` on 2026-02-22T12:20:35.000Z]*

*[Inline review on `packages/primeng/src/editor/editor.ts` could not be recreated]*

The `customModules` API is keyed by `Record<string, QuillModule>`, but it’s unclear from the docstring whether callers should pass the full Quill registration path (e.g. `modules/myModule`) or just the module name (e.g. `myModule`). Please clarify the expected key format (and whether the component auto-prefixes with `modules/`).
```suggestion
     * Map of custom Quill modules that need to be registered before initialization, see [here](https://quilljs.com/docs/modules#extending).
     * Keys must be full Quill registration paths (for example, `'modules/myModule'`), as they are passed directly to `Quill.register` and are not auto-prefixed by the component.
```

File: packages/primeng/src/editor/editor.ts [hunk 1]
```diff
@@ -291,14 +296,22 @@ export class Editor extends BaseEditableHolder<EditorPassThrough> {
...
                     this.dynamicQuill = quillModule.default;
+                    this.registerQuillModules();
                     this.createQuillEditor();
                 })
                 .catch((e) => console.error(e.message));
         } else {
+            this.registerQuillModules();
             this.createQuillEditor();
         }
     }
```
Comment (geromegrignon) [inline packages/primeng/src/editor/editor.ts:308]: *[Originally posted by `@Copilot` on 2026-02-22T12:20:35.000Z]*

New behavior registers custom Quill modules before initialization, but there are no corresponding unit tests. Since `packages/primeng/src/editor/editor.spec.ts` already covers other inputs (e.g. `modules`/`formats`), please add a test that asserts the provided `customModules` entries are registered on Quill before the editor instance is created.

**Comment by geromegrignon** [comment]
*[Originally posted by `@Copilot` on 2026-02-22T12:20:35.000Z]*

*[Inline review on `packages/primeng/src/types/editor/editor.types.ts` could not be recreated]*

Grammar: "Defined valid custom module." reads awkwardly. Consider changing it to "Defines a valid custom module." (or similar).
```suggestion
 * Defines a valid custom module.
```

**Comment by geromegrignon** [comment]
*[Originally posted by `@Copilot` on 2026-02-22T13:41:44.000Z]*

*[Inline review on `packages/primeng/src/editor/editor.spec.ts` could not be recreated]*

This assertion uses `toBe({})`, which checks reference equality and will always fail for object literals. Use deep equality (e.g., `toEqual({})`) to validate the default value instead.
```suggestion
            expect(editorInstance.customModules).toEqual({});
```

**Comment by geromegrignon** [comment]
*[Originally posted by `@Copilot` on 2026-02-22T13:41:45.000Z]*

*[Inline review on `packages/primeng/src/editor/editor.ts` could not be recreated]*

`Object.entries(this.customModules)` will throw if the input gets set to `undefined`/`null` at runtime (e.g., `[customModules]="maybeUndefined"`), overriding the default `{}`. Consider guarding (treat falsy/non-object as empty) before iterating so editor initialization doesn't crash.
```suggestion
        const modules = typeof this.customModules === 'object' && this.customModules !== null ? this.customModules : {};

        for (const [moduleName, moduleInstance] of Object.entries(modules)) {
```

File: packages/primeng/src/editor/editor.spec.ts [hunk 2]
```diff
@@ -547,6 +553,27 @@ describe('Editor', () => {
...
+            const registerSpy = jasmine.createSpy('register');
+
+            (editorInstance as any).dynamicQuill = {
+                register: registerSpy
+            };
+
+            (editorInstance as any).registerQuillModules();
+
+            expect(registerSpy).toHaveBeenCalledWith('modules/counter', component.customQuillModules['modules/counter']);
+        });
```
Comment (geromegrignon) [inline packages/primeng/src/editor/editor.spec.ts:576]: *[Originally posted by `@Copilot` on 2026-02-23T14:40:05.000Z]*

The test only verifies the happy path where custom modules are successfully registered. Consider adding test cases for edge cases such as: empty customModules object, invalid module definitions, or error scenarios where Quill.register might throw an exception. This would ensure the feature handles errors gracefully and doesn't break the editor initialization.
```suggestion
        });

        it('should handle empty custom quill modules without registering anything', () => {
            const editorEl = fixture.debugElement.query(By.css('p-editor'));
            const editorInstance = editorEl.componentInstance as Editor;
            const registerSpy = jasmine.createSpy('register');

            editorInstance.customModules = {};

            (editorInstance as any).dynamicQuill = {
                register: registerSpy
            };

            expect(() => (editorInstance as any).registerQuillModules()).not.toThrow();
            expect(registerSpy).not.toHaveBeenCalled();
        });

        it('should attempt to register invalid custom quill modules as provided', () => {
            const editorEl = fixture.debugElement.query(By.css('p-editor'));
            const editorInstance = editorEl.componentInstance as Editor;
            const registerSpy = jasmine.createSpy('register');

            editorInstance.customModules = {
                'modules/invalid': undefined as any
            };

            (editorInstance as any).dynamicQuill = {
                register: registerSpy
            };

            (editorInstance as any).registerQuillModules();

            expect(registerSpy).toHaveBeenCalledWith('modules/invalid', undefined as any);
        });

        it('should propagate errors thrown by quill register', () => {
            const editorEl = fixture.debugElement.query(By.css('p-editor'));
            const editorInstance = editorEl.componentInstance as Editor;

            editorInstance.customModules = {
                'modules/error': {} as any
            };

            (editorInstance as any).dynamicQuill = {
                register: () => {
                    throw new Error('register failed');
                }
            };

            expect(() => (editorInstance as any).registerQuillModules()).toThrowError('register failed');
        });
```

File: packages/primeng/src/editor/editor.ts [hunk 3]
```diff
@@ -291,14 +297,24 @@ export class Editor extends BaseEditableHolder<EditorPassThrough> {
...
 
+    private registerQuillModules(): void {
+        if (this.customModules) {
+            for (const [moduleName, moduleInstance] of Object.entries(this.customModules)) {
+                this.dynamicQuill.register(moduleName, moduleInstance);
```
Comment (geromegrignon) [inline packages/primeng/src/editor/editor.ts:313]: *[Originally posted by `@Copilot` on 2026-02-23T14:40:05.000Z]*

If `Quill.register()` throws an error for any of the custom modules (e.g., due to invalid module format or name collision), the entire registration loop will fail and subsequent modules won't be registered. Consider wrapping the register call in a try-catch block to log errors while allowing the editor to initialize with the successfully registered modules, or ensure that any registration failure is properly communicated to the user.
```suggestion
                try {
                    this.dynamicQuill.register(moduleName, moduleInstance);
                } catch (error: any) {
                    console.error(`Failed to register Quill module '${moduleName}':`, error);
                }
```

File: packages/primeng/src/types/editor/editor.types.ts [hunk 4]
```diff
@@ -208,3 +208,7 @@ export interface EditorTemplates {
      */
     header(): TemplateRef<any>;
 }
+/**
+ * Defines a valid custom module.
```
Comment (geromegrignon) [inline packages/primeng/src/types/editor/editor.types.ts:212]: *[Originally posted by `@Copilot` on 2026-02-23T14:40:05.000Z]*

The JSDoc comment for `QuillModule` is missing the `@group` tag that is consistently used for all other type definitions in this file (e.g., `@group Interface`, `@group Interfaces`, `@group Events`, `@group Templates`). Consider adding an appropriate group tag such as `@group Types` or `@group Interfaces` to maintain consistency with the documentation structure.
```suggestion
 * Defines a valid custom module.
 * @group Types
```

File: packages/primeng/src/editor/editor.ts [hunk 5]
```diff
@@ -291,14 +297,24 @@ export class Editor extends BaseEditableHolder<EditorPassThrough> {
...
+            }
+        }
+    }
```
Comment (geromegrignon) [inline packages/primeng/src/editor/editor.ts:316]: *[Originally posted by `@Copilot` on 2026-02-23T14:40:06.000Z]*

The `registerQuillModules()` method attempts to call `this.dynamicQuill.register()` without first checking if `this.dynamicQuill` is defined. While the current code paths always set `dynamicQuill` before calling this method, adding a null check would make the code more robust and prevent potential runtime errors if the initialization flow changes in the future. Consider adding a guard condition at the start of the method.

**Comment by geromegrignon** [comment]
Closed as too old.
If still relevant, a fresh new PR is welcome!
